### PR TITLE
[DOXIASITETOOLS-271] Overhaul locale support (make Locale#ROOT instead of Locale#ENGLISH default and use full locale)

### PR DESCRIPTION
**Pull request by michael-o**

### CODE REVIEW

**Comment by michael-o** [comment]
@hboutemy tested entire chain now with m-site-p, m-r-impl and m-project-info-reports-p. For me, this is good to be. Please have a look.

File: doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java [hunk 1]
```diff
@@ -63,8 +64,9 @@ Artifact getSkinArtifactFromRepository( ArtifactRepository localRepository,
      * Get a site descriptor from the project's site directory.
      *
      * @param siteDirectory the site directory, not null
-     * @param locale the locale wanted for the site descriptor. If not null, searching for
-     * <code>site_<i>localeLanguage</i>.xml</code>, otherwise searching for <code>site.xml</code>.
+     * @param locale the locale wanted for the site descriptor, not null. Telescoping lookup for
```
Comment (hboutemy) [inline doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java:67]: s/Telescoping/merging/

File: doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java [hunk 2]
```diff
@@ -63,8 +64,9 @@ Artifact getSkinArtifactFromRepository( ArtifactRepository localRepository,
...
+     * <code>site_language_country_variant.xml</code>, <code>site_language_country.xml</code>,
```
Comment (hboutemy) [inline doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java:68]: add "if provided locale defines a variant and/or a country and/or a language"

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site.xml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  <skin>
+    <groupId>org.apache.maven.skins</groupId>
+    <artifactId>maven-stylus-skin</artifactId>
+  </skin>
+  <body>
+    <links>
+      <item name="Maven 2" href="http://maven.apache.org/maven2/"/>
+    </links>
+
+    <menu name="Overview">
```
Comment (hboutemy) [inline doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site.xml:40]: for clarity, having "Overview from site.xml" would be useful

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  <skin>
+    <groupId>org.apache.maven.skins</groupId>
+    <artifactId>maven-stylus-skin</artifactId>
+  </skin>
+  <body>
+    <links>
+      <item name="Maven 2" href="http://maven.apache.org/maven2/"/>
+    </links>
+
+    <menu name="Overview">
```
Comment (hboutemy) [inline doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml:40]: for clarity, having "Overview from site_de_DE_BY.xml" would be useful

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  </bannerLeft>
+  <bannerRight>
+    <src>http://maven.apache.org/images/maven-small.gif</src>
+  </bannerRight>
+  <skin>
+    <groupId>org.apache.maven.skins</groupId>
+    <artifactId>maven-stylus-skin</artifactId>
+  </skin>
+  <body>
+    <links>
```
Comment (hboutemy) [inline doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml:36]: why override this link if it is not really changed?

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/language/src/site/site.xml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  <skin>
+    <groupId>org.apache.maven.skins</groupId>
+    <artifactId>maven-stylus-skin</artifactId>
+  </skin>
+  <body>
+    <links>
+      <item name="Maven 2" href="http://maven.apache.org/maven2/"/>
+    </links>
+
+    <menu name="Overview">
```
Comment (hboutemy) [inline doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/language/src/site/site.xml:40]: for clarity, having "Overview from site.xml" would be useful

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/language/src/site/site_de.xml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  <skin>
+    <groupId>org.apache.maven.skins</groupId>
+    <artifactId>maven-stylus-skin</artifactId>
+  </skin>
+  <body>
+    <links>
+      <item name="Maven 2" href="http://maven.apache.org/maven2/"/>
+    </links>
+
+    <menu name="Overview">
```
Comment (hboutemy) [inline doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/language/src/site/site_de.xml:40]: for clarity, having "Overview from site_de.xml" would be useful

**Review by hboutemy — CHANGES_REQUESTED** [review]
I'm not sure I fully understood how merging ("telescoping" if anybody understand this term) happens from most specific variant to region to language happens, bu I proposed to describe what I expect = an equivalent or resource bundles

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml (hunk 5 above)
Comment (michael-o) [reply doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml:36]: I didn't, I simply copied files to test SD discovery locally. This isn't an inheritance test.

**Comment by michael-o** [comment]
@hboutemy rewritten/applied. Please check again.

File: doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java [hunk 8]
```diff
@@ -63,8 +64,11 @@ Artifact getSkinArtifactFromRepository( ArtifactRepository localRepository,
      * Get a site descriptor from the project's site directory.
      *
      * @param siteDirectory the site directory, not null
-     * @param locale the locale wanted for the site descriptor. If not null, searching for
-     * <code>site_<i>localeLanguage</i>.xml</code>, otherwise searching for <code>site.xml</code>.
+     * @param locale the locale wanted for the site descriptor, not null. Most specific
```
Comment (hboutemy) [inline doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java:67]: the "most specific to least specific" part should be associated to the `@return` description

File: doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java [hunk 9]
```diff
@@ -63,8 +64,11 @@ Artifact getSkinArtifactFromRepository( ArtifactRepository localRepository,
...
+     * to least specific lookup from <code>site_language_country_variant.xml</code>,
+     * <code>site_language_country.xml</code>, <code>site_language.xml}</code>,
+     * to <code>site.xml</code> as last resort for {@link Locale#ROOT}. If provided
```
Comment (hboutemy) [inline doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/SiteTool.java:70]: s/./,/

File: doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml (hunk 5 above)
Comment (hboutemy) [reply doxia-integration-tools/src/test/resources/unit/site-tool-locales-test/full/src/site/site_de_DE_BY.xml:36]: notice that now I understand that there is no inheritance for locale-specific site descriptors: inheritance only happens against parent site descriptors...
For locale, it's only the most specific site descriptor that is loaded, there is no inheritance with less-specific ones

**Review by hboutemy — CHANGES_REQUESTED** [review]

**Comment by michael-o** [comment]
@hboutemy Fixed spots.

**Review by hboutemy — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #529

**Pull request merged**